### ISSUE TITLE
Failure to type check?

### ISSUE THREAD

**Issue by xogeny** [Microsoft/TypeScript#5688]
I was surprised to find the following code accepted by the TypeScript compiler since there is an obvious type error here:

```
type Update = (path: string, v: any) => void;

function foo() {
    var u: Update = (v: any): void => {
        console.log(v);
    }
}
```

Clearly, the `Update` type represents functions that take two arguments (an `string` and another value of type `any`).  But the variable `u` in the function `foo` is clearly assigned a value which is a function that takes only one argument.  This doesn't generate an error despite the fact that `u` has been explicitly declared as having the type `Update`.

Is there some reason that the TypeScript compiler doesn't catch such a basic type discrepancy?  Perhaps there is a good reason, but I can't figure out why this doesn't generate an error.

**Comment by subjectix** [comment]
The reason is **compatibility**. Typescript does a lot of work to ensure compatibility. The signature of the provided function `(v: any) => void` is compatible with the signature `(path: string, v: any) => void` since your **implementation** only cares about the first argument, and it takes `any` (which includes string) so if anyone calls `u` with the signature of `Update` the implementation is not broken. The vise versa is not valid though. You can't call a function that takes 2 arguments with only 1. This is a one-way compatibility. And it's not a bug, you simply **chose** not to care about the second argument. Just like event handlers, the signature might be `(e: Event, v: Value, t: Target) => void` but you only care about when the event is fired, so you just write: `() => {/*...*/}`. if the compiler was to complain about that you had to write `(e,v,t) => {/*...*/}` only to get it to compile, which is ugly. was that clear enough? :D

**Comment by ahejlsberg** [comment]
Correct, this is by design. The function assignment compatibility rules allow you to omit parameters you're not interested in (which is fairly common in JavaScript).
